### PR TITLE
BUG/ENH: `Styler.format()` always inherits from `.set_na_rep()`

### LINKED ISSUE #40032: BUG: Styler format `na_rep` arg overwrites existing when not given.
`Styler` has its own `format()` method that sets a formatter for columns, and a function `set_na_rep()` for dealing with missing data. 
It seems the two do not play well together:

```
df = pd.DataFrame([[1,2],[3,4]])
df.iloc[0,0] = np.nan
df.style.set_na_rep('MISS').set_precision(1)

MISS | 2
3.0 | 4

df.style.set_na_rep('MISS').format('{:,.1f}')

nan | 2.0
3.0 | 4.0

df.style.set_na_rep('MISS').format('{:,.1f}', na_rep='PASS')

PASS | 2.0
3.0 | 4.0
```

could probably do with a review how the attribute `na_rep` interacts with the formatter and try to make it more dynamic or use pointers instead of static value, if possible.

### PR DESCRIPTION
**Pull request by attack68**
- [X] closes #40032 
- [X] tests added: `test_display_format_subset_interaction`: why: see below.
- [X] tests removed: `test_format_bad_na_rep`: why: `na_rep` works if not string in some cases - changed to duck-typing.
- [X] tests lefts unchanged that pass: `test_display_format`, `test_display_set_precision`, `test_format_subset`, `test_display_dict`
- [X] tests expanded: `test_display_format_raises`
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

The issue is that there are currently two competing methods in `Styler` for missing value represenatation: 

  - `.set_na_rep('x')` and
  -  `.format(na_rep='y')`. 

The two do not interact well together dynamically. This PR brings the `format wrapping` into the class so that `format` can be updated and inherit values in `set_na_rep` after it has been called. The image below highlights the impact this PR has and the added test replicates this pattern:

Screen Shot 2021-02-25 at 20 38 45

Also at the bottom here, we add a `KeyError` to remove the ambiguity between what happens when a dict is provided with a column Key that is not specifically included within the provided `subset` slice.

### CODE REVIEW

File: pandas/io/formats/style.py [hunk 1]
```diff
@@ -575,75 +564,99 @@ def _translate(self):
 
         return d
 
-    def format(self, formatter, subset=None, na_rep: Optional[str] = None) -> Styler:
+    def format(
+        self,
+        formatter: Optional[
+            Union[Dict[Any, Optional[Union[str, Callable]]], str, Callable]
```
Comment (jreback) [inline pandas/io/formats/style.py:570]: can you type this at the top / in conftest.py (note that we can probably generally type these formatters across IO functions, but that's for another day)

File: pandas/io/formats/style.py [hunk 2]
```diff
@@ -575,75 +564,99 @@ def _translate(self):
...
+    ) -> Styler:
         """
         Format the text display value of cells.
 
         Parameters
         ----------
         formatter : str, callable, dict or None
-            If ``formatter`` is None, the default formatter is used.
+            Format specification to use for displaying values. If ``None``, the default
+            formatter is used. If ``dict``, keys should corresponcd to column names,
```
Comment (jreback) [inline pandas/io/formats/style.py:582]: correspond

File: pandas/io/formats/style.py [hunk 3]
```diff
@@ -575,75 +564,99 @@ def _translate(self):
...
+            and values should be string or callable.
         subset : IndexSlice
             An argument to ``DataFrame.loc`` that restricts which elements
             ``formatter`` is applied to.
         na_rep : str, optional
-            Representation for missing values.
-            If ``na_rep`` is None, no special formatting is applied.
+            Representation for missing values. If ``None``, will revert to using
+            ``Styler.na_rep``
```
Comment (jreback) [inline pandas/io/formats/style.py:589]: hmm why are you adding this special function? we don't do this anywhere else

File: pandas/io/formats/style.py [hunk 4]
```diff
@@ -1294,6 +1307,40 @@ def hide_columns(self, subset) -> Styler:
         self.hidden_columns = self.columns.get_indexer_for(hidden_df.columns)
         return self
 
+    def _default_formatter(self, x):
```
Comment (jreback) [inline pandas/io/formats/style.py:1310]: why is this not a module level function?

File: pandas/io/formats/style.py [hunk 5]
```diff
@@ -1294,6 +1307,40 @@ def hide_columns(self, subset) -> Styler:
...
+        if isinstance(x, (float, complex)):
+            return f"{x:.{self.precision}f}"
+        return x
+
+    def _maybe_wrap_formatter(
```
Comment (jreback) [inline pandas/io/formats/style.py:1315]: why is this moved? what is the point of self.na_rep?

**Review by jreback — CHANGES_REQUESTED** [review]
a number of questions

**Comment by jreback** [comment]
> .set_na_rep('x') and
.format(na_rep='y').

if there are 2 let's deprecate one. we use `.format()` everywhere else, I don't really see having a specific method na rep as all that useful (we *also* have set_precision as well).

we need one canoncial way to do things.

**Comment by jreback** [comment]
try not to mix bug fixes with api changes in a single pr. keep them as focused as possible.

File: pandas/io/formats/style.py (hunk 5 above)
Comment (attack68) [reply pandas/io/formats/style.py:1315]: its moved because as a module level it can't act dynamically and update `self.na_rep` if it is later changed. At class level it can so this does to inheritance.

File: pandas/io/formats/style.py (hunk 3 above)
Comment (attack68) [reply pandas/io/formats/style.py:589]: I think your comment about the fact that nowhere else is there a `set_na_rep` method that lends itself to the answer here. It is possible for there to be a solution where both functions (that currently exists) operate consistently, as opposed to somewhat inconsistently and inconveniently at present

Comment (attack68) [reply pandas/io/formats/style.py:589]: I would add that this might not even be interpreted as a change. `no special formatting applied` might be expected to be that `Styler.na_rep`` is used as default, it is ambiguous. I was just clarifying what the bug fix did here.

**Comment by attack68** [comment]
@jreback I agree somewhat that maybe multiple different `set_na_rep` and `set_precision` and `format` methods are not ideal and maybe incorporating everything to a format function is preferable. I've never suggested a deprecation previously (and all of these are legacy decisions from my perspective) but am happy to work on a solution otherwise, if you think thats better.

**Comment by attack68** [comment]
I think the PR above is better both in terms of API and code maintenance. Do close this PR if you agree.
